pkg/apis/coordination/validation/validation.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/kubernetes/pkg/apis/coordination"
+)
+
+// ValidateLease validates a Lease.
+func ValidateLease(lease *coordination.Lease) field.ErrorList {
+	allErrs := ValidateLeaseSpec(&lease.Spec, field.NewPath("spec"))
+	return allErrs
+}
+
+// ValidateLeaseUpdate validates an update of Lease object.
+func ValidateLeaseUpdate(lease, oldLease *coordination.Lease) field.ErrorList {
+	allErrs := ValidateLease(lease)
+	allErrs = append(allErrs, ValidateLeaseSpecUpdate(&lease.Spec, &oldLease.Spec, field.NewPath("spec"))...)
+	return allErrs
+}
+
+// ValidateLeaseSpec validates spec of Lease.
+func ValidateLeaseSpec(spec *coordination.LeaseSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if spec.LeaseDurationSeconds <= 0 {
+		fld := fldPath.Child("leaseDurationSeconds")
+		allErrs = append(allErrs, field.Invalid(fld, spec.LeaseDurationSeconds, "must be greater than 0"))
+	}
+	if spec.RenewTime.Before(&spec.AcquireTime) {
+		fld := fldPath.Child("renewTime")
+		allErrs = append(allErrs, field.Invalid(fld, spec.RenewTime, "must be greater than spec.acquireTime"))
+	}
+	if spec.LeaseTransitions < 0 {
+		fld := fldPath.Child("leaseTransitions")
+		allErrs = append(allErrs, field.Invalid(fld, spec.LeaseTransitions, "must to greater or equal than 0"))
+	}
+	return allErrs
+}
+
+// ValidateLeaseSpecUpdate validates an update of spec of Lease object.
+func ValidateLeaseSpecUpdate(spec, oldSpec *coordination.LeaseSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if spec.AcquireTime.Before(&oldSpec.AcquireTime) {
+		fld := fldPath.Child("acquireTime")
+		allErrs = append(allErrs, field.Invalid(fld, spec.AcquireTime, "can't go back in time"))
+	}
+	if spec.RenewTime.Before(&oldSpec.RenewTime) {
+		fld := fldPath.Child("renewTime")
+		allErrs = append(allErrs, field.Invalid(fld, spec.AcquireTime, "can't go back in time"))
+	}
+	if spec.LeaseTransitions < oldSpec.LeaseTransitions {
+		fld := fldPath.Child("leaseTransitions")
+		allErrs = append(allErrs, field.Invalid(fld, spec.LeaseTransitions, "can't be smaller"))
+	}
+	if spec.LeaseTransitions > oldSpec.LeaseTransitions+1 {
```

Comment:
consider requiring `LeaseTransitions` to increment if `HolderIdentity` changes

Reply:
Actually after thinking about it, I think we should allow for incrementing it even if holder didn't change. It's client reponsibility to set it, and I can imagine that in some cases we may want to reflect the fact that we are "re-acquiring" a lease not "renewing" it (e.g. component was restarted, lease expired in the meantime, but wasn't picked up by anyone - so now, we are effectively acquiring a lease so it's a transition).

Reply:
I agree that we should allow it to increment for the same identity.
Question is whether we should force it to increment for a new identity.

If we're trying to avoid clients *having* to reason about the fields, I guess that suggests the answer is no. Clients could still optionally do that if they want, though.

For new use cases (esp. Kubelet heartbeats, where identity doesn't change), clients could even choose to ignore everything but `HolderIdentity`, `LeaseDurationSeconds`, and `RenewTime`.

In practice, I think trying to make a distinction between `AcquireTime` and `RenewTime` when we don't have to, and when `RenewTime` is the true trigger for observers, will just end up being confusing. (But we have to do this for legacy things, IIUC)

Reply:
> If we're trying to avoid clients having to reason about the fields, I guess that suggests the answer is no. Clients could still optionally do that if they want, though.

+1

> In practice, I think trying to make a distinction between AcquireTime and RenewTime when we don't have to, and when RenewTime is the true trigger for observers, will just end up being confusing. (But we have to do this for legacy things, IIUC)

Yes - we need this for legacy reasons.
Also e.g. our leader-election library will be setting all fields for easier debugging. But we shouldn't force everyone to do that.

Reply:
>For new use cases (esp. Kubelet heartbeats, where identity doesn't change), clients could even choose to ignore everything but HolderIdentity, LeaseDurationSeconds, and RenewTime.

I think this is what I'd like to do for Kubelet heartbeats.